src/middlewared/middlewared/plugins/init_shutdown_script.py

```diff
@@ -104,33 +107,80 @@ class Config:
         if verrors:
             raise verrors
 
+    async def execute_task(self, task):
+        task_type = task['type']
+        cmd = None
+
+        if task_type == 'COMMAND':
+            cmd = task['command']
+        elif os.path.exists(task['script'] or '') and os.access(task['script'], os.X_OK):
+            cmd = f'exec {task["script"]}'
+
+        if cmd:
+            proc = await Popen(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                shell=True,
+                close_fds=True
+            )
+            await proc.wait()
+
+            if proc.returncode:
+                self.middleware.logger.debug(
+                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+                )
+
     @private
     @accepts(
         Str('when')
     )
-    def execute_init_tasks(self, when):
-        tasks = self.middleware.call_sync(
+    @job()
+    async def execute_init_tasks(self, job, when):
+
+        tasks = await self.middleware.call(
             'initshutdownscript.query', [
                 ['enabled', '=', True],
                 ['when', '=', when]
             ])
 
         for task in tasks:
-            task_type = task['type']
-            ret = None
-
-            if task_type == 'COMMAND':
-                ret = subprocess.run(
-                    task['command'],
-                    shell=True
-                )
-            elif os.path.exists(task['script']) and os.access(task['script'], os.X_OK):
-                ret = subprocess.run(
-                    f'exec {task["script"]}',
-                    shell=True
+            try:
+                await asyncio.wait_for(self.execute_task(task), timeout=task['timeout'])
+            except asyncio.TimeoutError:
+                self.middleware.logger.debug(
+                    f'{task["type"]} {task["command"] if task["type"] == "COMMAND" else task["script"]} timed out'
                 )
 
-            if ret and ret.returncode:
-                self.middleware.logger.debug(
-                    f'Execution failed for {task_type} {task["command"] if task_type == "COMMAND" else task["script"]}'
+        job.set_progress(100, f'Completed tasks for {when}')
+
+
+async def __event_system(middleware, event_type, args):
+    if args['id'] in ('shutdown', 'reboot'):
+        # Default 90 seconds is being added to scripts timeout to ensure other
+        # system related scripts can execute safely within the default timeout
+        timeout_value = 90 + sum(
```

Comment:
Default value for this sysctl is `kern.init_shutdown_timeout: 120`
And we should probably not change it if tunable exists.

Reply:
Updated :)